### PR TITLE
Clean up XCodec and other codecs

### PR DESCRIPTION
**Pull request by ebezzam**
# What does this PR do?

- Remove `**kwargs` input from methods
- Inherit from `PreTrainedAudioTokenizerBase` like DAC
- Remove creation of `self._input_length` inside `encode`, and use same truncation logic as in DAC's `forward`

### CODE REVIEW

File: src/transformers/models/dac/feature_extraction_dac.py [hunk 1]
```diff
@@ -150,10 +150,11 @@ def __call__(
             max_length=max_length,
             truncation=truncation,
             padding=padding,
-            return_attention_mask=False,
+            return_attention_mask=padding,
             pad_to_multiple_of=self.hop_length,
         )
-
+        if padding:
+            padded_inputs["padding_mask"] = padded_inputs.pop("attention_mask")
```
Comment (ebezzam) [inline src/transformers/models/dac/feature_extraction_dac.py:157]: Encodec feature extracture returns this. May be useful for post-processing? (removing padding during batch processing)

File: src/transformers/models/encodec/modeling_encodec.py [hunk 2]
```diff
@@ -449,7 +449,7 @@ def decode(self, codes: torch.Tensor) -> torch.Tensor:
 
 
 @auto_docstring
-class EncodecPreTrainedModel(PreTrainedModel):
+class EncodecPreTrainedModel(PreTrainedAudioTokenizerBase):
```
Comment (ebezzam) [inline src/transformers/models/encodec/modeling_encodec.py:452]: DAC using [it](https://github.com/huggingface/transformers/blob/7f2f53424ee18a8317f4bde93b579a3a8958690c/src/transformers/modeling_utils.py#L6282), so why not Encodec as well for standardizing

File: src/transformers/models/dac/feature_extraction_dac.py (hunk 1 above)
Comment (vasqu) [reply src/transformers/models/dac/feature_extraction_dac.py:157]: Ah hmm, I didn't want to introduce until we manage to prove that dac batching actually works :cry:

File: src/transformers/models/dac/modeling_dac.py [hunk 3]
```diff
@@ -613,6 +613,8 @@ def decode(
             The codebook indices for each codebook, representing the quantized discrete
             representation of the input. This parameter should be provided if you want
             to decode directly from the audio codes (it will overwrite quantized_representation).
+        return_dict (`bool`, *optional*, defaults to `True`):
+            Whether to return a [`DacDecoderOutput`] instead of a plain tuple.
```
Comment (vasqu) [inline src/transformers/models/dac/modeling_dac.py:617]: Does `auto_docstring` not catch it?

File: src/transformers/models/encodec/modeling_encodec.py (hunk 2 above)
Comment (vasqu) [reply src/transformers/models/encodec/modeling_encodec.py:452]: Love it! The purpose was to mark it as cross dep, e.g. dac for dia and encodec for musicgen iirc - needs a lot of refactoring in the future but the idea was to move things into the preprocessor instead of the modeling code (which always had to overwrite inits/frompretrained)

File: src/transformers/models/xcodec/modeling_xcodec.py [hunk 4]
```diff
@@ -446,15 +444,9 @@ def encode(
...
-        if input_values.ndim != 3:
-            raise ValueError(
-                f"Expected input shape (batch_size, channels, num_samples), but got shape {input_values.shape}"
-            )
-
-        _, channels, self._input_length = input_values.shape
-
-        if channels not in (1, 2):
-            raise ValueError(f"Number of audio channels must be 1 or 2, but got {channels}")
+        _, channels, _ = input_values.shape
```
Comment (vasqu) [inline src/transformers/models/xcodec/modeling_xcodec.py]:
```suggestion
        channels = input_values.shape[1]
```

File: src/transformers/models/xcodec/modeling_xcodec.py [hunk 5]
```diff
@@ -507,13 +501,6 @@ def decode(
         quantized_acoustic = self.fc2(quantized.transpose(1, 2)).transpose(1, 2)
         audio_values = self.acoustic_decoder(quantized_acoustic)
 
-        if getattr(self, "_input_length", None) is not None:
-            output_length = audio_values.shape[-1]
-            if self._input_length != output_length:
-                extra = output_length - self._input_length
-                start = extra // 2
-                audio_values = audio_values[..., start : start + self._input_length]
```
Comment (vasqu) [inline src/transformers/models/xcodec/modeling_xcodec.py:515]: A manual padding?! Just why was this there

**Review by vasqu — APPROVED** [review]
Looks good overall, can you run the slow tests just to be on the safe side

**Comment by ebezzam** [comment]
run-slow: dac, encodec, xcodec

File: src/transformers/models/xcodec/modeling_xcodec.py (hunk 5 above)
Comment (ebezzam) [reply src/transformers/models/xcodec/modeling_xcodec.py:515]: Because of this hack inside `encode`: https://github.com/huggingface/transformers/blob/586ec1fc8586ad7a0784c7c5ea6896cb0338ad04/src/transformers/models/xcodec/modeling_xcodec.py#L450 😅

File: src/transformers/models/dac/modeling_dac.py (hunk 3 above)
Comment (yaswanth19) [reply src/transformers/models/dac/modeling_dac.py:617]: Should use @can_return_tuple dec here

**Comment by ebezzam** [comment]
run-slow: dac, encodec, xcodec

**Comment by ebezzam** [comment]
run-slow: dac

File: src/transformers/models/xcodec/modeling_xcodec.py (hunk 5 above)
Comment (Manalelaidouni) [reply src/transformers/models/xcodec/modeling_xcodec.py:515]: No, this is to remove the padding introduced the original xcodec, so the decoded length exactly matches the original input (codec convention),
I've mentioned this early in the pr msg "Note that I removed the extra padding at the end of the decoded audio_values so that the output length matches that of the input, the produced audio closely resembles the original audio with a 1e-5 tolerance; otherwise the forward pass is identical with the output from the original model."

Comment (ebezzam) [reply src/transformers/models/xcodec/modeling_xcodec.py:515]: Yep it is helpful to remove the extra padding due to decoder. 

This is now handled in [forward](https://github.com/huggingface/transformers/blob/main/src/transformers/models/xcodec/modeling_xcodec.py#L548), in the same was as [DAC](https://github.com/huggingface/transformers/blob/main/src/transformers/models/dac/modeling_dac.py#L671).

So it does mean that if a user manually calls `encode` and `decode`, they will have to also manually remove the padding. But in general we shouldn't create new class variables inside the method

Comment (Manalelaidouni) [reply src/transformers/models/xcodec/modeling_xcodec.py:515]: Yup, understood
